packages/next/src/build/webpack/plugins/build-manifest-plugin.ts

```diff
@@ -273,15 +273,88 @@ export default class BuildManifestPlugin {
         if (SYSTEM_ENTRYPOINTS.has(entrypoint.name)) continue
         const pagePath = getRouteFromEntrypoint(entrypoint.name)
 
-        if (!pagePath) {
-          continue
-        }
+        if (!pagePath) continue
 
         const filesForPage = getEntrypointFiles(entrypoint)
-
         assetMap.pages[pagePath] = [...new Set([...mainFiles, ...filesForPage])]
       }
 
+      // When a CSS file is server-rendered, it is cleaned up during
+      // the client render. During navigation, if the target route
+      // dynamically imports the same CSS file, the client will skip
+      // loading it as it was already loaded statically, but still
+      // cleaned up during the client render. This results in missing
+      // the expected CSS styles.
+      // x-ref: `next/src/client/index.tsx`
+
+      // 1. page transition
+      // 2. import (CSS already exists)
+      // 3. client render starts
+      // 4. clean up server-rendered CSS
+      // 5. CSS is missing
+
+      // To prevent this, we add the dynamically imported CSS files
+      // to the build manifest to signal the client not to clean up
+      // the CSS file if it's loaded dynamically, preserving the styles.
+
+      /* Inspired by `next/src/build/webpack/react-loadable-plugin.ts` */
+      const _handleBlock = (block: any) => {
+        block.blocks.forEach(_handleBlock)
+
+        const dynamicCSSFiles = new Set<string>()
```

Comment:
If you collect dynamicCSSFiles here, will it also collect the css from sync dependencies, I should have a test for it that is not included in the final build manifest. Or al least not duplicated